node/hyperbahn/handler.js

```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
     var self = this;
     self.channel.topChannel.handler.removeServicePeer(service.serviceName, service.hostPort);
 };
+
+function convertHosts(hosts) {
+    var res = [];
+    for (var i = 0; i < hosts.length; i++) {
+        var strs = hosts[i].split(':');
+        var obj = {
+            port: parseInt(strs[1])
+        };
+        strs = strs[0].replace('localhost', '127.0.0.1').split('.');
+        obj.ip = {
+            ipv4: parseInt(strs[3]) + (parseInt(strs[2]) << 8) +
+                (parseInt(strs[1]) << 16) + (parseInt(strs[0]) << 24)
+        };
+
+        res.push(obj);
+    }
+
+    return res;
+}
+
+HyperbahnHandler.prototype.discover =
+function discover(self, req, head, body, cb) {
+    var serviceName = body.query.serviceName;
+    if (serviceName.length === 0) {
+        cb(null, {
+            ok: false,
+            body: InvalidServiceName({
+                serviceName: serviceName
+            }),
+            typeName: 'invalidServiceName'
+        });
+        return;
+    }
+
+    var exitNodes = self.egressNodes.exitsFor(serviceName);
+    var exitHosts = Object.keys(exitNodes);
+
+    var myHost = self.channel.hostPort;
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
+        self.sendRelayAsThrift({
+            hostPort: exitHosts[exit],
+            body: body,
+            inreq: req,
+            endpoint: 'Hyperbahn::discover'
+        }, cb);
```

Comment:
Instead of passing in `cb` you want to do something like

function handleForward(err, resp) {
    if (resp) {
        return cb(null, resp)
    }

    logAThingy();
    sendErrorFrame(errors.classify(err), err.message)
}
```